### PR TITLE
Fix extract from stream

### PR DESCRIPTION
**Pull request by misthophoroi**
Changes:

fix unexpected behavior:
SevenZipExtractor dispose archiveStream when created by public SevenZipExtractor(Stream archiveStream)

fix perfomance degradation at SevenZipBase.GetUniqueID()
under high load Identifiers list's length increases upto 80000 and when new object created all cpu time spended at SevenZipBase.GetUniqueID()

add net462 target - fupported long paths by UNC (\\\\?\disk:\folder\file)

fixed netstandard2.0 target in test projects

### CODE REVIEW

File: SevenZip/ArchiveEmulationStreamProxy.cs [hunk 1]
```diff
@@ -18,16 +18,19 @@ internal class ArchiveEmulationStreamProxy : Stream, IDisposable
...
         public Stream Source { get; }
 
+        bool _dispose = true;
         /// <summary>
         /// Initializes a new instance of the ArchiveEmulationStream class.
         /// </summary>
         /// <param name="stream">The stream to wrap.</param>
         /// <param name="offset">The stream offset.</param>
-        public ArchiveEmulationStreamProxy(Stream stream, int offset)
+        /// <param name="dispose">Dispose wraped stream</param>
```
Comment (vasily-kirichenko) [inline SevenZip/ArchiveEmulationStreamProxy.cs:27]: typo: wrapped

Comment (vasily-kirichenko) [reply SevenZip/ArchiveEmulationStreamProxy.cs:27]: Maybe name it `liveOpen`, analogues to some other .net API?

File: SevenZip/ArchiveEmulationStreamProxy.cs [hunk 2]
```diff
@@ -18,16 +18,19 @@ internal class ArchiveEmulationStreamProxy : Stream, IDisposable
         /// </summary>
         public Stream Source { get; }
 
+        bool _dispose = true;
```
Comment (vasily-kirichenko) [inline SevenZip/ArchiveEmulationStreamProxy.cs:21]: `readonly`?

File: SevenZip/SevenZipBase.cs [hunk 3]
```diff
@@ -15,7 +15,8 @@ public abstract class SevenZipBase : MarshalByRefObject
     {
         private readonly bool _reportErrors;
         private readonly int _uniqueID;
-        private static readonly List<int> Identifiers = new List<int>();
+        private static readonly object _sync_uniqueIds = new object();
```
Comment (vasily-kirichenko) [inline SevenZip/SevenZipBase.cs:18]: Please, use camelCase (`_syncUniqueIds`)

File: SevenZip/SevenZipBase.cs [hunk 4]
```diff
@@ -132,18 +121,6 @@ protected SevenZipBase(string password = "")
...
-        /// </summary>
-        ~SevenZipBase()
-        {
-            // This lock probably isn't necessary but just in case...
-            lock (Identifiers)
-            {
-                Identifiers.Remove(_uniqueID);
-            }
-        }
-
```
Comment (vasily-kirichenko) [inline SevenZip/SevenZipBase.cs:146]: Great!

File: SevenZip/SevenZipExtractor.cs [hunk 5]
```diff
@@ -107,7 +107,7 @@ private void Init(Stream stream)
             SevenZipLibraryManager.LoadLibrary(this, _format);
             try
             {
-                _inStream = new ArchiveEmulationStreamProxy(stream, _offset);
+                _inStream = new ArchiveEmulationStreamProxy(stream, _offset, false);
```
Comment (vasily-kirichenko) [inline SevenZip/SevenZipExtractor.cs:110]: Maybe make the last parameter optional (`dispose = false`) to make the API backward compatible?

File: SevenZip/ArchiveEmulationStreamProxy.cs (hunk 2 above)
Comment (misthophoroi) [reply SevenZip/ArchiveEmulationStreamProxy.cs]: ok

File: SevenZip/SevenZipBase.cs (hunk 3 above)
Comment (misthophoroi) [reply SevenZip/SevenZipBase.cs]: ok

File: SevenZip/ArchiveEmulationStreamProxy.cs [hunk 6]
```diff
@@ -72,12 +75,14 @@ public override void Write(byte[] buffer, int offset, int count)
 
         public new void Dispose()
         {
-            Source.Dispose();
+            if(_leaveOpen)
```
Comment (vasily-kirichenko) [inline SevenZip/ArchiveEmulationStreamProxy.cs:78]: Shouldn't it be `if (!_leaveOpen)`?

File: SevenZip/ArchiveEmulationStreamProxy.cs [hunk 7]
```diff
@@ -72,12 +75,14 @@ public override void Write(byte[] buffer, int offset, int count)
...
+                Source.Dispose();
         }
 
         public override void Close()
         {
-            Source.Close();
+            if(_leaveOpen)
```
Comment (vasily-kirichenko) [inline SevenZip/ArchiveEmulationStreamProxy.cs:84]: Same as above.

File: SevenZip/SevenZipExtractor.cs [hunk 8]
```diff
@@ -107,7 +107,7 @@ private void Init(Stream stream)
             SevenZipLibraryManager.LoadLibrary(this, _format);
             try
             {
-                _inStream = new ArchiveEmulationStreamProxy(stream, _offset);
+                _inStream = new ArchiveEmulationStreamProxy(stream, _offset, leaveOpen: true);
```
Comment (vasily-kirichenko) [inline SevenZip/SevenZipExtractor.cs:110]: `leaveOpen: false`?

File: SevenZip/SevenZipExtractor.cs [hunk 9]
```diff
@@ -124,7 +124,7 @@ private void Init(Stream stream)
                     _format = InArchiveFormat.PE;
                     try
                     {
-                        _inStream = new ArchiveEmulationStreamProxy(stream, _offset);
+                        _inStream = new ArchiveEmulationStreamProxy(stream, _offset, leaveOpen: true);
```
Comment (vasily-kirichenko) [inline SevenZip/SevenZipExtractor.cs:127]: Same as above.

File: SevenZip/SevenZip.csproj [hunk 10]
```diff
@@ -94,5 +94,6 @@
     <Reference Include="System.Configuration" Condition="'$(TargetFramework)' == 'net45'">
       <HintPath>C:\Windows\Microsoft.NET\assembly\GAC_MSIL\System.Configuration\v4.0_4.0.0.0__b03f5f7f11d50a3a\System.Configuration.dll</HintPath>
     </Reference>
+    <Reference Include="System.Configuration" Condition="'$(TargetFramework)' == 'net462'" />
```
Comment (vasily-kirichenko) [inline SevenZip/SevenZip.csproj:97]: I'm not sure if there should be `<HintPath>` here. Should it?

**Comment by squid-box** [comment]
Sorry for the silence, this was quite a big PR and a little hard to follow. 

In general I appreciate the effort and it seems mostly fine, but I'll add some questions as review comments.

File: .github/workflows/dotnetcore.yml [hunk 11]
```diff
@@ -0,0 +1,35 @@
+name: .NET Core
```
Comment (squid-box) [inline .github/workflows/dotnetcore.yml:1]: The inclusion of a workflows config is a bit surprising. What is the reasoning for this?

File: SevenZip.Tests/SevenZip.Tests.csproj [hunk 12]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net45</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.1;net45;net462</TargetFrameworks>
```
Comment (squid-box) [inline SevenZip.Tests/SevenZip.Tests.csproj:3]: Why is this target changed from .NET Standard 2.0 to Core 2.1, while the SevenZip project itself remains Standard 2.0?

File: SevenZip.sln [hunk 13]
```diff
@@ -1,40 +1,40 @@
-﻿
```
Comment (squid-box) [inline SevenZip.sln:1]: Not really a problem, but a little odd to include what I guess is whitespace changes in a PR 🤷

File: SevenZip/ArchiveExtractCallback.cs [hunk 14]
```diff
@@ -564,6 +564,11 @@ private static void ValidateFileNameAndCreateDirectory(string fileName)
         /// <returns></returns>
         private static string RemoveIllegalCharacters(string str, bool isDirectory = false)
         {
+#if NET462
+            var isUncPath = str.StartsWith(@"\\?\");
+            if(isUncPath)
+                str = str.Substring(@"\\?\".Length);
```
Comment (squid-box) [inline SevenZip/ArchiveExtractCallback.cs:570]: Also only preference, but I'd really prefer

```
if (isUncPath)
{
    str = str.Substring(@"\\?\".Length);
}
```

Applies to more places in this PR.

File: SevenZip/ArchiveExtractCallback.cs [hunk 15]
```diff
@@ -564,6 +564,11 @@ private static void ValidateFileNameAndCreateDirectory(string fileName)
         /// <returns></returns>
         private static string RemoveIllegalCharacters(string str, bool isDirectory = false)
         {
+#if NET462
+            var isUncPath = str.StartsWith(@"\\?\");
```
Comment (squid-box) [inline SevenZip/ArchiveExtractCallback.cs:568]: I would make `@"\\?\"` a variable that can then be accesses in the other places in this function.

File: SevenZip/SevenZip.csproj (hunk 10 above)
Comment (squid-box) [reply SevenZip/SevenZip.csproj:97]: I would actually merge the two instead:

`<Reference Include="System.Configuration" Condition="'$(TargetFramework)' == 'net45' OR '$(TargetFramework)' == 'net462'" />`

And yes, HintPath can be removed.

File: SevenZip/ArchiveEmulationStreamProxy.cs [hunk 16]
```diff
@@ -18,16 +18,19 @@ internal class ArchiveEmulationStreamProxy : Stream, IDisposable
         /// </summary>
         public Stream Source { get; }
 
+        readonly bool _leaveOpen = false;
```
Comment (squid-box) [inline SevenZip/ArchiveEmulationStreamProxy.cs:21]: Once again, purely preference but please add a new line below this line to separate it from the following comment :)

File: SevenZip/SevenZipExtractor.cs [hunk 17]
```diff
@@ -135,17 +136,18 @@ private void Init(Stream stream)
...
+
         /// <summary>
         /// Initializes a new instance of SevenZipExtractor class.
         /// </summary>
         /// <param name="archiveStream">The stream to read the archive from.
+        /// <param name="leaveOpen">true to leave the wraped stream open after the ArchiveEmulationStreamProxy object is disposed; otherwise, false.</param>
         /// Use SevenZipExtractor(string) to extract from disk, though it is not necessary.</param>
         /// <remarks>The archive format is guessed by the signature.</remarks>
-        public SevenZipExtractor(Stream archiveStream)
+        public SevenZipExtractor(Stream archiveStream, bool leaveOpen = false)
```
Comment (squid-box) [inline SevenZip/SevenZipExtractor.cs:148]: Pretty much just a suggestion, but wouldn't it be better to reverse this parameter - `bool disposeStream = true`?

Since the use of leaveOpen downstream is always `if  (!leaveOpen)` it would make for easier reading of the code, in my opinion at least.
The param comment would then be something like `Whether or not to dispose the stream after operation is complete.`.

File: SevenZip/SevenZipBase.cs [hunk 18]
```diff
@@ -15,7 +15,8 @@ public abstract class SevenZipBase : MarshalByRefObject
     {
         private readonly bool _reportErrors;
         private readonly int _uniqueID;
-        private static readonly List<int> Identifiers = new List<int>();
+        private static readonly object _syncUniqueIds = new object();
```
Comment (squid-box) [inline SevenZip/SevenZipBase.cs:18]: Once again preference, but renaming this something like `_uniqueIdLock` would make its purpose/use easier to identify.

File: SevenZip/SevenZipBase.cs (hunk 4 above)
Comment (squid-box) [reply SevenZip/SevenZipBase.cs:146]: I'm not 100% sure what the use of UniqueID is, while this implementation seems faster, it's also limited to ~4 million instances where the previous implementation was not

Can you shed some light on this?

File: .github/workflows/dotnetcore.yml (hunk 11 above)
Comment (misthophoroi) [reply .github/workflows/dotnetcore.yml:1]: I have add it to find out why some of SevenZipCompressorTests.cs tests failed at AppVeyor
but then I fix tests at a5d0dd4. If you do not plan to use github workflow for tests can delete it

File: SevenZip.Tests/SevenZip.Tests.csproj (hunk 12 above)
Comment (misthophoroi) [reply SevenZip.Tests/SevenZip.Tests.csproj:3]: it is only one way how I can run tests with not net45 and net462 frameworks
I found it at https://github.com/nunit/docs/wiki/.NET-Core-and-.NET-Standard#why-cant-my-tests-target-net-standard

File: SevenZip.sln (hunk 13 above)
Comment (misthophoroi) [reply SevenZip.sln:1]: ok. I will fix it

File: SevenZip/SevenZipExtractor.cs (hunk 17 above)
Comment (misthophoroi) [reply SevenZip/SevenZipExtractor.cs:148]: I have change it after @vasily-kirichenko review to make it more like [System.IO.StreamReader](https://docs.microsoft.com/en-US/dotnet/api/system.io.streamreader.-ctor?view=netcore-3.1#System_IO_StreamReader__ctor_System_IO_Stream_System_Text_Encoding_System_Boolean_System_Int32_System_Boolean_)

File: SevenZip/SevenZipBase.cs (hunk 18 above)
Comment (misthophoroi) [reply SevenZip/SevenZipBase.cs]: fixed

File: SevenZip/ArchiveEmulationStreamProxy.cs (hunk 16 above)
Comment (misthophoroi) [reply SevenZip/ArchiveEmulationStreamProxy.cs:21]: fixed

File: SevenZip/ArchiveExtractCallback.cs (hunk 15 above)
Comment (misthophoroi) [reply SevenZip/ArchiveExtractCallback.cs]: fixed

File: SevenZip/ArchiveExtractCallback.cs (hunk 14 above)
Comment (misthophoroi) [reply SevenZip/ArchiveExtractCallback.cs]: fixed

File: SevenZip/ArchiveEmulationStreamProxy.cs (hunk 6 above)
Comment (misthophoroi) [reply SevenZip/ArchiveEmulationStreamProxy.cs]: fixed

File: SevenZip/ArchiveEmulationStreamProxy.cs (hunk 7 above)
Comment (misthophoroi) [reply SevenZip/ArchiveEmulationStreamProxy.cs]: fixed

File: SevenZip/SevenZip.csproj (hunk 10 above)
Comment (misthophoroi) [reply SevenZip/SevenZip.csproj]: merged and removed
